src/java.base/share/classes/com/sun/crypto/provider/AESCipher.java

```diff
@@ -641,5 +650,127 @@ protected void engineUpdateAAD(ByteBuffer src) {
+            return 0;
+        }
+        int outLenNeeded = engineGetOutputSize(inLen);
+
+        if (output.remaining() < outLenNeeded) {
+            throw new ShortBufferException("Need at least " + outLenNeeded
+                + " bytes of space in output buffer");
+        }
+
+        // detecting input and output buffer overlap may be tricky
+        // we can only write directly into output buffer when we
+        // are 100% sure it's safe to do so
+
+        boolean a1 = input.hasArray();
+        boolean a2 = output.hasArray();
+        int total = 0;
+
+        if (a1) { // input has an accessible byte[]
+            byte[] inArray = input.array();
+            int inOfs = input.arrayOffset() + inPos;
+
+            if (a2) { // output has an accessible byte[]
+                byte[] outArray = output.array();
+                int outPos = output.position();
+                int outOfs = output.arrayOffset() + outPos;
+
+                // check array address and offsets and use temp output buffer
+                // if output offset is larger than input offset and
+                // falls within the range of input data
+                boolean useTempOut = false;
+                if (inArray == outArray &&
+                    ((inOfs < outOfs) && (outOfs < inOfs + inLen))) {
+                    useTempOut = true;
+                    outArray = new byte[outLenNeeded];
+                    outOfs = 0;
+                }
+                if (isUpdate) {
+                    total = engineUpdate(inArray, inOfs, inLen, outArray, outOfs);
+                } else {
+                    total = engineDoFinal(inArray, inOfs, inLen, outArray, outOfs);
+                }
+                if (useTempOut) {
+                    output.put(outArray, outOfs, total);
+                } else {
+                    // adjust output position manually
+                    output.position(outPos + total);
+                }
+                // adjust input position manually
+                input.position(inLimit);
+            } else { // output does not have an accessible byte[]
+                byte[] outArray;
+                if (isUpdate) {
+                    outArray = engineUpdate(inArray, inOfs, inLen);
+                } else {
+                    outArray = engineDoFinal(inArray, inOfs, inLen);
+                }
+                if (outArray != null && outArray.length != 0) {
+                    output.put(outArray);
+                    total = outArray.length;
+                }
+                // adjust input position manually
+                input.position(inLimit);
+            }
+        } else { // input does not have an accessible byte[]
+            if (core.getMode() == CipherCore.GCM_MODE) {
+                if (isUpdate) {
+                    return core.update(input, output);
+                } else {
+                    return core.doFinal(input, output);
+                }
```

Comment:
I didn't see a way to override this because CipherSpi is a public class, any methods I added would become a new API. 
 Also bufferCrypt is private, so I had to copy it.  CipherSpi does not know which mode is being used, but AESCipher does.  Maybe I'm missing something, I'd rather it not be a copy, but I couldn't see a better way. If you have a specific idea, please give me details.

Reply:
How about overriding
protected int engineDoFinal(ByteBuffer input, ByteBuffer output) throws ... {
    if (core.getMode() == CipherCore.GCM_MODE && !input.hasArray()) {
        // call your own optimized byte buffer code path
    } else {
        super.engineDoFinal(input, output);
    }
}

Would this work? If yes, then no need to duplicate the whole bufferCrypt() method.

Reply:
Good call.. I wouldn't have thought about super.